In [1]:
import sys
import os

project_root = os.path.abspath('..')
if project_root not in sys.path:
    sys.path.append(project_root)
 
%load_ext autoreload
%autoreload 2

In [2]:
import pandas as pd
import numpy as np
import polars as pl

import matplotlib.pyplot as plt
import seaborn as sns

import warnings

from ydata_profiling import ProfileReport
import src.data.make_dataset as data
import glob
import json

warnings.filterwarnings("ignore")
sns.set_style()
plt.rcParams["figure.figsize"] = (12, 6)


In [3]:
df_raw = pd.concat(
    (
        pd.read_excel(archivo, usecols=data.COLS_A_OCUPAR.keys())
        for archivo in glob.glob("../data/raw/diagnosticos/*.xlsx")
    )
)
df_raw = df_raw.drop(columns=data.COLS_A_ELIMINAR + ["Rut Paciente", "Rut Profesional"])

In [4]:
df_raw["Código Diagnóstico"] = df_raw["Código Diagnóstico"].astype(str)

In [5]:
df_limpia = pd.read_csv(
    "../data/processed/datos_limpios_diagnosticos.csv", encoding="latin-1", sep=";"
)


# Analisis exploratorio previo al preprocesamiento

In [6]:
report = ProfileReport(df_raw)
report.to_file("../reports/1.0-jrb-exploratory-data-analysis.html")


Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

## Analisis de Columna Codigo Diagnostico

In [7]:
largo_codigos = df_raw["Código Diagnóstico"].str.len().value_counts(normalize=True, dropna=False)
display(largo_codigos * 100)

5    58.598147
3    27.038000
7     9.194450
2     1.983357
6     1.550353
4     1.053433
1     0.581900
8     0.000360
Name: Código Diagnóstico, dtype: float64

- El 58.6% de los codigos presenta largo 5. Los codigos con largo 5 son del tipo: "C38.4" o "C34_1", por
lo que es el formato deseable

In [8]:
df_raw[df_raw["Código Diagnóstico"].str.len() == 5]["Código Diagnóstico"].sample(10)

61564     J67.8
198580    G12.2
140407    I25.5
259254    I35.0
137754    I27.0
105795    D38.1
194795    J44.9
13861     G47.3
245951    J98.6
361510    C34_9
Name: Código Diagnóstico, dtype: object


- El 27.0% de los codigos presenta largo 3. Los codigos con largo 3 son del tipo: "R05", "C34", por
lo que corresponde a la Categoria del diagnostico.


In [9]:
largo_3 = df_raw[df_raw["Código Diagnóstico"].str.len() == 3][
    ["Código Diagnóstico", "Nombre Diagnóstico"]
].value_counts()
display(largo_3)


Código Diagnóstico  Nombre Diagnóstico                                               
I99                 Otros trastornos y los no especificados del sistema circulatorio     29603
J47                 Bronquiectasia                                                       15305
I50                 Insuficiencia cardíaca                                               13381
C34                 Tumor maligno de los bronquios y del pulmón                          11886
J84                 Otras enfermedades pulmonares intersticiales                         11185
                                                                                         ...  
I13                 Hipertensión cardíaca y enfermedad renal.                                1
R96                 Otras muertes súbitas de causa desconocida                               1
H60                 Otitis externa                                                           1
H55                 Nistagmo y otros movimientos oculares i

- El 9.19% de los codigos presenta largo 7. Los codigos con largo 7 son del tipo: "J44.8.4" o "E10-E16".

In [10]:
largo_7 = df_raw[df_raw["Código Diagnóstico"].str.len() == 7]["Código Diagnóstico"].sample(10)
display(largo_7)

160630    I20-I25
245583    J45.0.2
280180    J44.8.4
218799    J45.0.2
148900    J45.9.1
315657    R90-R94
103099    J45.0.1
98049     E10-E16
74955     J45.0.1
325137    J45.0.1
Name: Código Diagnóstico, dtype: object

- El 1.98% de los codigos presenta largo 2. Los codigos con largo 2 son del tipo: "55", por lo que
tienen una notacion de diagnostico deficiente. Sin embargo, estos codigos se relacionan a un
unico "Nombre Diagnostico". Esto puede permitir recodificar estos codigos de forma mas facil.

In [11]:
largo_2 = df_raw[df_raw["Código Diagnóstico"].str.len() == 2][
    ["Código Diagnóstico", "Nombre Diagnóstico"]
].value_counts()
display(largo_2)


Código Diagnóstico  Nombre Diagnóstico                    
55                  Fibrosis pulmonar                         4335
79                  Trasplante  Pulmon                        2788
78                  Trasplante Cardiaco                       1197
41                  Nódulo pulmonar                            842
20                  Bronquiectasias                            363
54                  Enfermedad pulmonar difusa                 342
42                  Tumor pulmonar                             113
77                  Asma Bronquial en el Adulto (Sospecha)     110
74                  Tabaquismo                                 100
64                  Hipertensión pulmonar primaria              99
17                  Neumonitis por hipersensibilidad            98
49                  Disnea en estudio                           79
44                  Cáncer pulmonar                             53
21                  Bronquiectasias infectadas                  50
59 

- El 1.55% de los codigos presenta largo 6. Los codigos con largo 6 son del tipo: "300-23". Tambien tienen
una relacion 1:1 entre Código Diagnóstico y Nombre Diagnóstico.

In [12]:
largo_6 = df_raw[df_raw["Código Diagnóstico"].str.len() == 6][
    ["Código Diagnóstico", "Nombre Diagnóstico"]
].value_counts()
display(largo_6)


Código Diagnóstico  Nombre Diagnóstico                                                             
300-23              Tetralogía de Fallot (TOF)                                                         2523
300-32              Ventriculo Unico Funcional                                                         1385
300-28              Transposición Grandes Arterias (discordancia de la conexión ventriculoarterial)     848
300-29              Transposición Grandes Arterias + CIV                                                519
300-12              Doble Salida VD (DSVD)                                                              517
                                                                                                       ... 
902-12              Preexcitación ventricular                                                             2
I51.14              CIV Sub Aórtica (tipo fallot)                                                         2
902-27              Macroreentrada p

- El 1.05% de los codigos presenta largo 4 y son del tipo: "81-4" o "I32*"

In [13]:
largo_4 = df_raw[df_raw["Código Diagnóstico"].str.len() == 4][
    ["Código Diagnóstico", "Nombre Diagnóstico"]
].value_counts()
display(largo_4)


Código Diagnóstico  Nombre Diagnóstico                                     
81-4                Otro                                                       4553
88-1                Otro:                                                       300
81-6                Estenosis Pulmonar Valvular                                 254
91-3                Enfermedad Coronaria 3V                                     251
85-6                Cardiomiopatía Isquemica                                    126
91-1                Enfermedad Coronaria 1V                                      80
85-3                Enfermedad Coronaria (cardiopatía coronaria)                 73
91-2                Enfermedad Coronaria 2V                                      66
91-5                Aneurisma Aorta Ascendnte (sin mención de ruptura)           43
91-6                Aneurisma Raíz Aórtica (sin mención de ruptura)              25
81-5                Estenosis Pulmonar Subvalvular (Infundibular)                16


- El 0.58% de los codigos presenta largo 1, y son del tipo "5" o "6"

In [14]:
largo_1 = df_raw[df_raw["Código Diagnóstico"].str.len() == 1][
    ["Código Diagnóstico", "Nombre Diagnóstico"]
].value_counts()
display(largo_1)


Código Diagnóstico  Nombre Diagnóstico          
5                   Otro                            3072
6                   Ex TBC pulmonar                  107
2                   Crisis asmática                   23
7                   TBC pulmonar                      23
8                   TBC pulmonar multiresistente       5
9                   Micobacteriosis atipica            2
dtype: int64

- El 0.00036% de los codigos presenta largo 8, y son del tipo "Z00- Z13"

In [15]:
largo_8 = df_raw[df_raw["Código Diagnóstico"].str.len() == 8][
    ["Código Diagnóstico", "Nombre Diagnóstico"]
].value_counts()
display(largo_8)


Código Diagnóstico  Nombre Diagnóstico                                  
Z00- Z13            PRUEBAS PARA ACLARAR O INVESTIGAR PROBLEMAS DE SALUD    2
dtype: int64

## Identificacion de codigos con mas de una glosa

In [35]:
duplas_codigo_glosa_diagnostico = (
    df_raw[["Código Diagnóstico", "Nombre Diagnóstico"]].value_counts().reset_index(name="conteo")
)

codigos_con_mas_de_una_glosa = duplas_codigo_glosa_diagnostico[
    duplas_codigo_glosa_diagnostico.duplicated("Código Diagnóstico", keep=False)
].sort_values("Código Diagnóstico")

In [37]:
display(codigos_con_mas_de_una_glosa)

Código Diagnóstico                                 Nombre Diagnóstico  \
115            J44.8.1                    Bronquitis crónica con enfisema   
489            J44.8.1  Bronquitis crónica con enfisema_x000D_\n_x000D...   
15             J45.0.3                                       Asma atópica   
129            J45.0.3                                     Asma atópica\t   
33             J45.1.1                                   Asma no alérgica   
166            J45.1.1                          Asma no alérgica_x000D_\n   

     conteo  
115     763  
489      49  
15     6922  
129     621  
33     4250  
166     377

- Por lo tanto, la base presenta 3 codigos con mas de una glosa (J44.8.1, J45.0.3, J45.1.1) entre
2016 e inicio de 2023. Esto sugiere que asignar un CIE genera una glosa especifica.

## Preprocesamiento de Diagnostico

- Teniendo en cuenta los resultados previos, una propuesta tentativa para el preprocesamiento
de los codigos de diagnosticos es:

    1. Eliminar el simbolo "."
    2. Eliminar los espacios
    2. A los que tienen largo 3, agregar una X al final hasta tener un largo 4
    3. A los que tienen largo 7, se separaran en diags con largo 5 (ej: J4591) y largo 7 (ej: E65-E68).
    Los que sigan teniendo largo 7 corresponderan a los Capitulos, por lo que sera necesario observar
    que hacer con esos diagnosticos (ya que son muy poco especificos)
    4. A los que tienen largo 2, 6, 4 y 1 se puede crear un diccionario para transformar el codigo a un
    codigo cie

# Analisis de base preprocesada

## Analisis de Columna Codigo Diagnostico

In [16]:
largos = df_limpia["codigo_diagnostico"].str.len().value_counts().index
descripcion_largos = (
    df_limpia["codigo_diagnostico"].str.len().value_counts(normalize=True, dropna=False)
)
display(descripcion_largos * 100)

4    97.275477
7     1.451923
6     0.561707
2     0.439646
5     0.251280
1     0.019967
Name: codigo_diagnostico, dtype: float64

Luego del preprocesamiento, los porcentajes de tipos de codigos fueron los siguientes:

- El 58.6% de los codigos de la base tienen largo 4
- El 27.4% de los codigos de la base tienen largo 3
- El 9.0% de los codigos de la base tienen largo 5

Esto ya abarca el 95% de toda la base de datos

In [17]:
resultados = {}
for largo in largos:
    filtro_por_largo = df_limpia["codigo_diagnostico"].str.len() == largo

    resumen_diags_con_largo = df_limpia[filtro_por_largo][
        ["codigo_diagnostico", "nombre_diagnostico"]
    ].value_counts()

    resultados[f"largo_{largo}"] = resumen_diags_con_largo

    print(f"Los codigos con largo {largo} tienen la siguiente forma:")
    print(resumen_diags_con_largo.sample(30, replace=True).index.get_level_values(0))
    print()


Los codigos con largo 4 tienen la siguiente forma:
Index(['K26X', 'Z453', 'I31X', 'F09X', 'I361', 'B48X', 'S81X', 'R31X', 'I712',
       'L022', 'A184', 'D64X', 'G500', 'E040', 'I455', 'J43X', 'C56X', 'L20X',
       'Q323', 'C838', 'G46X', 'T455', 'I839', 'I368', 'D481', 'C37X', 'Q259',
       'C637', 'F512', 'L024'],
      dtype='object', name='codigo_diagnostico')

Los codigos con largo 7 tienen la siguiente forma:
Index(['L00-L08', 'N17-N19', 'G80-G83', 'S00-S09', 'J85-J86', 'I30-I52',
       'L50-L54', 'E70-E90', 'G90-G99', 'R10-R19', 'E10-E16', 'L00-L08',
       'K40-K46', 'J60-J70', 'Q20-Q28', 'G80-G83', 'J00-J06', 'C40-C41',
       'Q30-Q34', 'D80-D89', 'H15-H22', 'E00-E07', 'G80-G83', 'I10-I15',
       'M80-M94', 'J10-J18', 'G60-G64', 'D37-D48', 'C00-C08', 'G20-G26'],
      dtype='object', name='codigo_diagnostico')

Los codigos con largo 6 tienen la siguiente forma:
Index(['400-22', '400-14', '902-14', '904-10', '200-17', '902-12', '902-28',
       '400-29', '400-15', '900-11'

In [18]:
with pd.ExcelWriter("../data/interim/diagnosticos_encontrados.xlsx") as file:
    for nombre_archivo, df_a_guardar in resultados.items():
        df_a_guardar.to_excel(file, sheet_name=nombre_archivo)
